text/0020-unique-record-keys.md

```diff
@@ -0,0 +1,107 @@
+And finally, it's mitigated by the fact that Cedar's current behavior on duplicate keys was not documented in the public docs (e.g., in [this section](https://docs.cedarpolicy.com/syntax-datatypes.html#record)), so if any policies are relying on this behavior, it's probably by accident.
+
+## Alternatives
+
+The main alternative is doing nothing, which of course has the reverse pros/cons of the proposal as explained in "Motivation" and "Drawbacks".
+
+Another alternative, which would address the first motivation and not the second, would be to change the EST format so that it supports record literals with duplicate keys.
+I believe this would require a breaking change to the EST format, and record literals would not be expressible as JSON maps, which is unfortunate because I believe that's the most natural/ergonomic.
+(Devil's advocate: we could maybe find a way to support JSON maps in the common case with some escape syntax for expressing record literals that have duplicate keys. If we want to take this alternative seriously, I can investigate this.)
+
+## Unresolved questions
+
+### Order-independence for record keys
+At first glance, one would hope that this change would be sufficient to ensure that the order of keys in a record literal (or more generally, any record value) does not matter.
+However, that's not quite true -- if keys are mapped to expressions with errors, the order does matter, because only the first error encountered will be returned.
+While we're making this change, should we also make a change so that we have true order-independence for Cedar record keys?
+I could think of at least three ways to do that:
+1. Always evaluate all values in a record, and return potentially multiple evaluation errors rather than just the first
+2. Provide no guarantee about which error the user will receive when multiple record keys have errors (nondeterministic error behavior) -- I think the formal spec currently guarantees the first error?
+3. Guarantee that record values will be evaluated in lexicographic order of their keys, regardless of the order the keys appear in the policy text
+
+This issue is particularly hairy for records contained in attribute values in the entity store -- I believe these records go through a hashmap representation in our current datapath, so we lose ordering information and also duplicate keys (silently). Since the values in these records are restricted expressions, there are very few ways they can produce evaluation errors, but they still can, in the case of calls to extension functions. This is probably a potential source of difference between the errors produced in the Rust implementation and in the formal spec; I wonder why DRT hasn't caught this, particularly in the case where we drop a duplicate key that was mapped to an erroring expression. Maybe the Dafny JSON parser silently takes last-value-wins? or maybe we don't test extension values in attribute values?
```

Comment:
I believe that during DRT we pass in the same post-hashmap entity store & context to both the Dafny and Rust, which means attributes were already silently de-duped in both cases. 

After poking around a bit, it looks like the only place we can generate records with duplicates is when [generating `Expr`s to include in the policy](https://github.com/cedar-policy/cedar-spec/blob/3f3e25b86c4bcd8c67567923dfe57e90e2ceb767/cedar-policy-generators/src/schema.rs#L2972). I believe evaluation will return an error in both the [Rust](https://github.com/cedar-policy/cedar/blob/e940bdf48d750834b96c3279ebd4f96ff8b0eb3d/cedar-policy-core/src/evaluator.rs#L589) and [Dafny](https://github.com/cedar-policy/cedar-spec/blob/3f3e25b86c4bcd8c67567923dfe57e90e2ceb767/cedar-dafny/def/engine.dfy#L138) if any value in a record fails to evaluate (regardless of duplicate keys).